File: twomap.c
#include "twomap.h"
#include "twomap.r"

void *twomap_ctor(void *const _self, va_list *app)
{
    struct twomap_class_s *const self = _self;
    if(self->vt != twomap)
        return 0;
    self->map = va_arg(*app, char*);
    self->objs_on_map = NULL;
    self->num_of_objs = 0;
    self->max_objs = va_arg(*app, int);
    self->w = va_arg(*app, int);
    self->h = va_arg(*app, int);
    self->max_w = va_arg(*app, int);
    self->max_h = va_arg(*app, int);
    self->no = va_arg(*app, int);
      
        
    return _self;
}

void twomap_print(const void *const _self)
{
    printf("%s\n", ((const struct twomap_class_s *const)_self)->map);
}

struct twomap_vt_s twomap_vt = {
    TWOMAP_VT_INIT
};

void *const twomap = &twomap_vt;

File: twomap.r
#ifndef CMDSNK_TWOMAP_R_SENTRY
#define CMDSNK_TWOMAP_R_SENTRY

#include <object.r>

#define TWOMAP_VT           OBJECT_VT

#define TWOMAP_VT_INIT      OBJECT_VT_INIT,                     \
                            .ctor=twomap_ctor,                  \
                            .print=twomap_print,                \
                            .size=sizeof(struct twomap_class_s)

#define TWOMAP_DATA         OBJECT_DATA;            \
                            char *map;              \
                            void **objs_on_map;     \
                            int num_of_objs;        \
                            int max_objs;           \
                            int w, h, max_w, max_h; \
                            int no

struct twomap_vt_s {
    TWOMAP_VT;
}; 

struct twomap_class_s {
    struct twomap_vt_s *vt;
    TWOMAP_DATA;
};

/* 2D MAP SHOULD CONSTIST OF:
    * pointer to a map (in a form of string pointer?)
    * max width and hight
    * collection of objects on the map
    * number of objects on the map
    * max number of objects allowed on the map?
*/

#endif
